File: weather/formatters/cli_text.py
"""
CLI text formatter for weather data.

Formats weather data as plain text for terminal output.
"""

from ..models import WeatherData
from .base import WeatherFormatter


class CliTextFormatter(WeatherFormatter):
    """
    Plain text formatter for CLI output.

    Produces emoji-annotated weather reports suitable for terminal display.
    Only includes fields that have data present (no "N/A" placeholders).
    """

    @property
    def platform(self) -> str:
        """Target platform name."""
        return "text"

    def format_current(self, data: WeatherData) -> str:
        """
        Format current weather for CLI display.

        Args:
            data: Current weather data

        Returns:
            Plain text weather report
        """
        lines: list[str] = []

        lines.append(f"{data.emoji} Weather for {data.location}")
        lines.append(f"🌡️ Temperature: {data.temperature:.0f}°C")

        feels = data.effective_feels_like
        if abs(feels - data.temperature) > 0.5:
            lines.append(f"   Feels like: {feels:.0f}°C")

        if data.temp_high is not None and data.temp_low is not None:
            lines.append(f"   Range: {data.temp_low:.0f}° - {data.temp_high:.0f}°")

        if data.humidity is not None:
            lines.append(f"💧 Humidity: {data.humidity}%")

        wind = data.wind_str
        if wind != "N/A":
            lines.append(f"💨 Wind: {wind}")

        if data.precipitation_chance is not None:
            lines.append(f"🌧️ Rain chance: {data.precipitation_chance}%")

        if data.uv_index is not None:
            lines.append(f"☀️ UV Index: {data.uv_index}")

        if data.aqhi is not None:
            lines.append(f"🌫️ AQHI: {data.aqhi_str}")
        elif data.aqi is not None:
            lines.append(f"🌫️ AQI: {data.aqi_str}")

        lines.append(f"📍 Provider: {data.provider_name}")

        return "\n".join(lines)

    def format_forecast(self, data: list[WeatherData]) -> str:
        """
        Format multi-day forecast for CLI display.

        Args:
            data: List of forecast weather data

        Returns:
            Plain text forecast report
        """
        lines: list[str] = ["📊 Weather Forecast\n"]

        for day in data:
            if day.forecast_date:
                date_str = day.forecast_date.strftime("%Y-%m-%d")
            else:
                date_str = "Unknown"

            temp_high = f"{day.temp_high:.0f}°" if day.temp_high is not None else "?"
            temp_low = f"{day.temp_low:.0f}°" if day.temp_low is not None else "?"

            condition = day.description or day.condition_raw or str(day.condition.value)
            lines.append(f"{date_str}: {temp_high} / {temp_low} — {condition}")

        return "\n".join(lines)

File: src/formatters/cli_text.ts
/* eslint-disable @typescript-eslint/naming-convention */
/**
 * CLI text formatter for weather data.
 *
 * Port of `weather/formatters/cli_text.py`. Produces emoji-annotated
 * plain-text reports for terminal display. Only fields with data are
 * rendered — no "N/A" placeholders, matching the Python implementation.
 *
 * Platform string is **"text"** (matches Python's
 * `CliTextFormatter.platform`).
 */

import {
  aqhiStr,
  aqiStr,
  effectiveFeelsLike,
  getEmoji,
  windStr,
} from "../models.js";
import type { WeatherData } from "../types.js";
import {
  formatIsoDate,
  formatTempBare,
  formatTempC,
  formatUv,
  uvDescription,
} from "../utils.js";
import { BaseFormatter } from "./base.js";

export class CliTextFormatter extends BaseFormatter {
  get platform(): string {
    return "text";
  }

  protected formatCurrent(data: WeatherData): string {
    const lines: string[] = [];

    lines.push(`${getEmoji(data.condition)} Weather for ${data.location}`);
    lines.push(`🌡️ Temperature: ${formatTempC(data.temperature)}`);

    const feels = effectiveFeelsLike(data);
    if (Math.abs(feels - data.temperature) > 0.5) {
      lines.push(`   Feels like: ${formatTempC(feels)}`);
    }

    if (data.temp_high != null && data.temp_low != null) {
      lines.push(
        `   Range: ${formatTempBare(data.temp_low)} - ${formatTempBare(data.temp_high)}`,
      );
    }

    if (data.humidity != null) {
      lines.push(`💧 Humidity: ${data.humidity}%`);
    }

    const wind = windStr(data);
    if (wind !== "N/A") {
      lines.push(`💨 Wind: ${wind}`);
    }

    if (data.precipitation_chance != null) {
      lines.push(`🌧️ Rain chance: ${data.precipitation_chance}%`);
    }

    if (data.uv_index != null) {
      lines.push(
        `☀️ UV Index: ${formatUv(data.uv_index)} (${uvDescription(data.uv_index)})`,
      );
    }

    if (data.aqhi != null) {
      lines.push(`🌫️ AQHI: ${aqhiStr(data)}`);
    } else if (data.aqi != null) {
      lines.push(`🌫️ AQI: ${aqiStr(data)}`);
    }

    lines.push(`📍 Provider: ${data.provider_name}`);

    return lines.join("\n");
  }

  protected formatForecast(data: WeatherData[]): string {
    const lines: string[] = ["📊 Weather Forecast\n"];

    for (const day of data) {
      const dateStr = day.forecast_date ? formatIsoDate(day.forecast_date) : "Unknown";
      const tempHigh =
        day.temp_high != null ? formatTempBare(day.temp_high) : "?";
      const tempLow =
        day.temp_low != null ? formatTempBare(day.temp_low) : "?";

      // Match Python: `day.description or day.condition_raw or str(day.condition.value)`
      const condition =
        day.description || day.condition_raw || day.condition;
      lines.push(`${dateStr}: ${tempHigh} / ${tempLow} — ${condition}`);
    }

    return lines.join("\n");
  }
}
